File: day-21/allergen_assessment.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strings"
)

const inputFile string = "./day-21/input.txt"

func ParseInput(input string) (map[string]map[string]bool, map[string]int) {
	allergenMatches := make(map[string]map[string]bool)
	ingredientCounts := make(map[string]int)

	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		splits := strings.SplitN(line, " (contains ", 2)
		ingredients := strings.Split(splits[0], " ")
		allergensString := splits[1][:len(splits[1])-1]
		allergens := strings.Split(allergensString, ", ")

		for _, allergen := range allergens {
			if _, found := allergenMatches[allergen]; !found {
				allergenMatches[allergen] = make(map[string]bool)
				for _, ingredient := range ingredients {
					allergenMatches[allergen][ingredient] = true
				}
			} else {
				// set intersection
				for i := range allergenMatches[allergen] {
					contains := false
					for _, ingredient := range ingredients {
						if i == ingredient {
							contains = true
						}
					}
					if contains == false {
						delete(allergenMatches[allergen], i)
					}
				}
			}
		}
		for _, ingredient := range ingredients {
			if _, found := ingredientCounts[ingredient]; !found {
				ingredientCounts[ingredient] = 0
			}
			ingredientCounts[ingredient]++
		}
	}
	return allergenMatches, ingredientCounts
}

func Part1(input string) int {
	allergenMatches, ingredientCounts := ParseInput(input)

	count := 0
	for i, c := range ingredientCounts {
		candidateAllergen := false
		for _, is := range allergenMatches {
			if _, found := is[i]; found {
				candidateAllergen = true
			}
		}
		if !candidateAllergen {
			count += c
		}
	}
	return count
}

func Part2(input string) string {
	allergenMatches, _ := ParseInput(input)

	matchedAllergens := make(map[string]string)
	for len(allergenMatches) > 0 {
		for a, is := range allergenMatches {
			if len(is) == 1 {
				var ingredient string
				for i := range is {
					ingredient = i
				}
				matchedAllergens[a] = ingredient
				delete(allergenMatches, a)
				//Remove ingredient from other allergen sets
				for otherAl, _ := range allergenMatches {
					delete(allergenMatches[otherAl], ingredient)
				}
			}
		}
	}

	//Get list of ingredients, sorted by allergen
	allergens := make([]string, 0, len(matchedAllergens))
	for a := range matchedAllergens {
		allergens = append(allergens, a)
	}
	sort.Strings(allergens)

	ingredients := make([]string, 0, len(allergens))
	for _, a := range allergens {
		ingredients = append(ingredients, matchedAllergens[a])
	}

	return strings.Join(ingredients, ",")
}

func main() {
	b, err := ioutil.ReadFile(inputFile)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %s\n", Part2(data))
}

File: day-21/allergen_assessment.py
import re, collections

input_file = 'day-21/input.txt'

def parse_input(input):
  allergen_matches = {}
  ingredient_count = collections.defaultdict(int)

  for line in input.splitlines():
    ingredients, allergens = line.split('(contains', 1)
    ingredients = re.findall(r'\w+', ingredients)
    allergens = re.findall(r'\w+', allergens)
    for allergen in allergens:
      if allergen not in allergen_matches:
        allergen_matches[allergen] = set(ingredients)
      else:
        allergen_matches[allergen] &= set(ingredients) # &= does intersection
    for ingredient in ingredients:
      ingredient_count[ingredient] += 1
  return allergen_matches, ingredient_count

def part1(input):
  allergen_matches, ingredient_count = parse_input(input)
  candidate_allergens = set().union(*allergen_matches.values())
  return sum (count for ingred, count in ingredient_count.items() if ingred not in candidate_allergens)

def part2(input):
  allergen_matches, _ = parse_input(input)
  candidate_allergens = set().union(*allergen_matches.values())
  matched_allergens = {}

  while len(candidate_allergens) > 0:
    for allergen in sorted(allergen_matches.keys(), key=lambda k: len(allergen_matches[k])):
      allergen_matches[allergen] &= candidate_allergens
      if len(allergen_matches[allergen]) == 1:
        # Guranteed_match
        candidate_allergens ^= allergen_matches[allergen] # ^= does not intersection
        matched_allergens[allergen] = allergen_matches[allergen].pop()

  ordered_ingredients = []
  for allergen in sorted(matched_allergens.keys()):
    ordered_ingredients.append(matched_allergens[allergen])

  return ','.join(ordered_ingredients)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))